From 8e30a413651c5600ee8385abfd67be2f1b853a4d Mon Sep 17 00:00:00 2001
From: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Date: Wed, 12 Dec 2012 17:38:04 +0100
Subject: [PATCH 67/78] [PORT FROM JBMR0] wifi: call
 ieee80211_sched_scan_stopped on interface removal

BZ: 73498

Port from wl18xx (not yet backported to wl12xx R5)
The interface might go down before we got the SCHED_STOPPED event,
so make sure to call ieee80211_sched_scan_stopped() if the scanned interface
is removed.
Replace sched_scanning with sched_vif in order to save the scanned interface.

Change-Id: I14af59610cc0bc4a8bf9f63f449853e45757687a
Orig-Change-Id: Ided7f2d10aa466e42dec94ad82831fe1687453f6
Signed-off-by: Eliad Peller <eliad@wizery.com>
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Reviewed-on: http://android.intel.com:8080/85613
Reviewed-by: Naoumenko, PaulX <paulx.naoumenko@intel.com>
Reviewed-by: Bachot, Jean-michel <jean-michel.bachot@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 .../drivers/net/wireless/wl12xx/debugfs.c          |    1 -
 wl12xx-compat/drivers/net/wireless/wl12xx/event.c  |    4 ++--
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c   |   18 +++++++-----------
 wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h |    2 +-
 4 files changed, 10 insertions(+), 15 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/debugfs.c b/wl12xx-compat/drivers/net/wireless/wl12xx/debugfs.c
index 59cc5d9..283ca92 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/debugfs.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/debugfs.c
@@ -600,7 +600,6 @@ static ssize_t driver_state_read(struct file *file, char __user *user_buf,
 	DRIVER_STATE_PRINT_HEX(platform_quirks);
 	DRIVER_STATE_PRINT_HEX(chip.id);
 	DRIVER_STATE_PRINT_STR(chip.fw_ver_str);
-	DRIVER_STATE_PRINT_INT(sched_scanning);
 
 #undef DRIVER_STATE_PRINT_INT
 #undef DRIVER_STATE_PRINT_LONG
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/event.c b/wl12xx-compat/drivers/net/wireless/wl12xx/event.c
index f104236..edf3caa 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/event.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/event.c
@@ -130,9 +130,9 @@ static int wl1271_event_process(struct wl1271 *wl, struct event_mailbox *mbox)
 	if (vector & PERIODIC_SCAN_COMPLETE_EVENT_ID) {
 		wl1271_debug(DEBUG_EVENT, "PERIODIC_SCAN_COMPLETE_EVENT "
 			     "(status 0x%0x)", mbox->scheduled_scan_status);
-		if (wl->sched_scanning) {
+		if (wl->sched_vif) {
 			ieee80211_sched_scan_stopped(wl->hw);
-			wl->sched_scanning = false;
+			wl->sched_vif = NULL;
 		}
 	}
 
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 98406ed..429d4af 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -1470,11 +1470,6 @@ static void wl1271_recovery_work(struct work_struct *work)
 	/* Prevent spurious TX during FW restart */
 	ieee80211_stop_queues(wl->hw);
 
-	if (wl->sched_scanning) {
-		ieee80211_sched_scan_stopped(wl->hw);
-		wl->sched_scanning = false;
-	}
-
 	/* reboot the chipset */
 	while (!list_empty(&wl->wlvif_list)) {
 		wlvif = list_first_entry(&wl->wlvif_list,
@@ -2494,7 +2489,6 @@ static void wl1271_op_stop_locked(struct wl1271 *wl)
 	wl->time_offset = 0;
 	wl->ap_fw_ps_map = 0;
 	wl->ap_ps_map = 0;
-	wl->sched_scanning = false;
 	memset(wl->roles_map, 0, sizeof(wl->roles_map));
 	memset(wl->links_map, 0, sizeof(wl->links_map));
 	memset(wl->roc_map, 0, sizeof(wl->roc_map));
@@ -2925,6 +2919,11 @@ static void __wl1271_op_remove_interface(struct wl1271 *wl,
 		ieee80211_scan_completed(wl->hw, true);
 	}
 
+	if (wl->sched_vif == wlvif) {
+		ieee80211_sched_scan_stopped(wl->hw);
+		wl->sched_vif = NULL;
+	}
+
 	if (!test_bit(WL1271_FLAG_RECOVERY_IN_PROGRESS, &wl->flags)) {
 		/* disable active roles and clear RX filters */
 		ret = wl1271_ps_elp_wakeup(wl);
@@ -3199,10 +3198,8 @@ static int wl1271_sta_handle_idle(struct wl1271 *wl, struct wl12xx_vif *wlvif,
 		clear_bit(WLVIF_FLAG_IN_USE, &wlvif->flags);
 	} else {
 		/* The current firmware only supports sched_scan in idle */
-		if (wl->sched_scanning) {
+		if (wl->sched_vif == wlvif)
 			wl1271_scan_sched_scan_stop(wl, wlvif);
-			ieee80211_sched_scan_stopped(wl->hw);
-		}
 
 		ret = wl12xx_start_dev(wl, wlvif);
 		if (ret < 0)
@@ -4021,7 +4018,7 @@ static int wl1271_op_sched_scan_start(struct ieee80211_hw *hw,
 	if (ret < 0)
 		goto out_sleep;
 
-	wl->sched_scanning = true;
+	wl->sched_vif = wlvif;
 
 out_sleep:
 	wl1271_ps_elp_sleep(wl);
@@ -6314,7 +6311,6 @@ static struct ieee80211_hw *wl1271_alloc_hw(void)
 	wl->ap_fw_ps_map = 0;
 	wl->quirks = 0;
 	wl->platform_quirks = 0;
-	wl->sched_scanning = false;
 	wl->system_hlid = WL12XX_SYSTEM_HLID;
 	wl->active_sta_count = 0;
 	wl->fwlog_size = 0;
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h b/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
index 3f52f87..355c746 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
@@ -475,7 +475,7 @@ struct wl1271 {
 	struct wl1271_scan scan;
 	struct delayed_work scan_complete_work;
 
-	bool sched_scanning;
+	struct wl12xx_vif *sched_vif;
 
 	/* The current band */
 	enum ieee80211_band band;
-- 
1.7.9

